tap_agent/src/tap/account.rs

```diff
@@ -0,0 +1,846 @@
+            r#"
+            WITH rav AS (
+                SELECT 
+                    rav -> 'message' ->> 'timestamp_ns' AS timestamp_ns 
+                FROM 
+                    scalar_tap_latest_ravs 
+                WHERE 
+                    allocation_id = $1 
+                    AND sender_address = $2
+            ) 
+            SELECT 
+                COUNT(*), 
+                MAX(id), 
+                SUM(value) 
+            FROM 
+                scalar_tap_receipts 
+            WHERE 
+                allocation_id = $1 
+                AND sender_address = $2 
+                AND CASE WHEN (
+                    SELECT 
+                        timestamp_ns :: NUMERIC 
+                    FROM 
+                        rav
+                ) IS NOT NULL THEN timestamp_ns > (
+                    SELECT 
+                        timestamp_ns :: NUMERIC 
+                    FROM 
+                        rav
+                ) ELSE TRUE END
+            "#,
+            inner
+                .allocation_id
+                .to_string()
+                .trim_start_matches("0x")
+                .to_owned(),
+            inner.sender.to_string().trim_start_matches("0x").to_owned()
+        )
+        .fetch_one(&inner.pgpool)
+        .await?;
+
+        let mut unaggregated_fees = inner.unaggregated_fees.lock().await;
+
+        // `COUNT(*)` will always return a value, so we don't need to check for `None`.
+        match res.count.unwrap() {
+            0 => {
+                unaggregated_fees.last_id = 0;
+                unaggregated_fees.value = 0;
+            }
+            // If the count is non-zero, then `MAX(id)` and `SUM(value)` will be non-null.
+            // If they are null, then something is extremely wrong with the database.
+            _ => {
+                ensure!(
+                    res.max.is_some(),
+                    "MAX(id) is null but the receipt COUNT(*) is not zero"
+                );
+                ensure!(
+                    res.sum.is_some(),
+                    "SUM(value) is null, but the receipt COUNT(*) is not zero"
+                );
+
+                unaggregated_fees.last_id = res.max.unwrap().try_into()?;
+                unaggregated_fees.value = res.sum.unwrap().to_string().parse::<u128>()?;
+            }
+        }
+
+        // TODO: check if we need to run a RAV request here.
+
+        Ok(())
+    }
+
+    /// Request a RAV from the sender's TAP aggregator.
+    /// Will remove the aggregated receipts from the database if successful.
+    async fn rav_requester(inner: Arc<Inner>) {
+        Self::rav_requester_try(&inner).await.unwrap_or_else(|e| {
+            error!(
+                "Error while requesting a RAV for allocation {} and sender {}: {:?}",
+                inner.allocation_id, inner.sender, e
+            );
+        });
+    }
+
+    async fn rav_requester_try(inner: &Arc<Inner>) -> anyhow::Result<()> {
+        loop {
```

Comment:
Related to this: do we make sure that we don't request multiple RAVs for the same receipts in parallel if a new receipt comes in before the RAV comes back and we still have unaggregated fees > threshold?

Reply:
Yes because we always (unless I missed it somewhere) check `Self::rav_requester_task_is_running` before starting the `rav_requester` in an async thread.

Reply:
- We do make sure that we don't do RAV requests in parallel for any given (sender,allocation), by checking `Self::rav_requester_task_is_running` before spawning a RAV request task. Could be made more elegant I think.
- Each RAV request must contain the last RAV (unless it is the first RAV request). We basically aggregate the receipts into the latest RAV we got, and we get a new RAV out of that. That's why it is very important the the whole RAV request dance is serial.
- The TapManager will only aggregate receipts that have a timestamp more recent than the last RAV (though note that we are OK with multiple receipts having the same timestamp, they also have a nonce). To make sure that there is no confusion with the new receipts coming in *while* we're requesting a RAV, the TapManager will only aggregate receipts up to timestamps that are more than `timestamp_buffer` old: https://github.com/semiotic-ai/timeline-aggregation-protocol/blob/656fc15700f6d29299d78f9c790539cbddb68d02/tap_core/src/tap_manager/manager.rs#L179

  One issue I can see here is that we are not checking (in indexer-service) that new receipts have a reasonably accurate timestamp (let's say within 5 seconds of our own clock). So as it is right now it is not really safe :/.

Reply:
Yup, created and issue: #92